### ISSUE TITLE
Memory leak in Python client when calling Channel.subscribe()

### ISSUE THREAD

**Issue by kevin-wright** [grpc/grpc#9461]
There appears to be a memory leak when using the grpcio Python client and calling Channel.subscribe().  The leak is pretty small (roughly 5MB to 6MB per hour), but it does add up over time.

I'm seeing this when using Python 2.7 on Ubuntu 14.04 (64-bit).  I see the same behavior in grpcio versions 1.0.0 and 1.0.4.

This issue is easy to reproduce using the Python "helloworld" example, with a few small changes.  The steps to reproduce are:

1. git clone https://github.com/grpc/grpc.git

2. cd into examples/python/helloworld

3. Change the contents of greeter_client.py so that instead of using the GreeterStub, we call channel.subscribe():

```
from __future__ import print_function

import grpc
import time

#import helloworld_pb2
#import helloworld_pb2_grpc


def run():
  channel = grpc.insecure_channel('localhost:50051')
  #stub = helloworld_pb2_grpc.GreeterStub(channel)
  #response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
  #print("Greeter client received: " + response.message)

  channel.subscribe(on_connectivity_change, try_to_connect=True)

  while True:
    time.sleep(60)

  return

def on_connectivity_change(value):
  print("connection changed: " + str(value))
  return

if __name__ == '__main__':
  run()
```

4. Run the server: `python greeter_server.py`

5. Run the client: `python greeter_client.py`

6. Note the client's process ID.

7. Track the client process's virtual memory usage and resident memory (I just used `top -p <proces_id>`).  The resident memory will grow at ~264KiB every 3 minutes (~5MiB per hour).  Eventually, once the resident memory reaches a certain point (after roughly 11.5 hours), the virtual memory will jump up by 64MiB.  Virtual memory will then remain constant while resident memory continues to increase at ~5MiB per hour.  After another ~11.5 hours, virtual memory will jump up by 64MiB again, and the process will repeat.  

This is a subtle bug to notice, since the leak is so slow, and virtual memory stays steady for such an extended period of time.  But after running for 5 days, the process will be using roughly half a gigabyte of memory, so it does affect long-running processes.  (Processes that are shorter running, or channels that are not long-lived, will probably not be materially affected.)

I've included a graph below of both the virtual memory usage and the resident memory.

![memory_usage](https://cloud.githubusercontent.com/assets/6549159/22267142/aa6b8756-e250-11e6-8aa2-307d03891587.png)

**Comment by nathanielmanistaatgoogle** [comment]
Very little changed between `1.0.4` and `1.1.0`, but please confirm that you are able to reproduce this issue with gRPC Python `1.1.0`?

**Comment by kevin-wright** [comment]
Yes, it appears that the same thing is happening in 1.1.0.  It's been running for a few hours now, and I'm seeing very similar behavior.  I'll let it run overnight and let you know for sure.  But the initial indications are that this issue is also in 1.1.0.

**Comment by kevin-wright** [comment]
So I left the test running for a few days, and I can confirm that we're still seeing the same behavior in 1.1.0.

**Comment by kevin-wright** [comment]
For what it's worth, we do seem to be seeing the same behavior in grpcio 1.2.1.

**Comment by baconalot** [comment]
We are also seeing this for grpcio 1.3.0.
This is free mem for a couple of servers. The only difference between the green one and the others is that the green one does not call `channel.subscribe`:
![image](https://cloud.githubusercontent.com/assets/1876368/25813177/e4721bd0-3418-11e7-9154-58351d6483e6.png)

**Comment by kpayson64** [comment]
This should be fixed in 1.3.5

**Comment by mwes** [comment]
FYI we retested under 1.4.0 and confirm this fixed the issue @kevin-wright was seeing. Thanks!
